hooks: pkg_resources: implement support for package content listing

rokm opened the pull request:
Implement a custom `pkg_resources` provider for our `FrozenImporter`, by subclassing a `pkg_resources.NullProvider` and overriding its `_has()`, `_isdir()`, and `_listdir()` methods. These are required/used by:
* `resource_exists()` function (or provider's `has_resource()` method)
* `resource_isdir()` function (or provider's `resource_isdir()` method)
* `resource_listdir()` function (or provider's `resource_listdir()` method)

and the default implementations provided by NullProvider raise `NotImplementedError("Can't perform this operation for unregistered loader type")`.

The implementation supports both external data files (the ones collected into the frozen application's directory, e.g. from wheels) and embedded data files (collected into PYZ archive, e.g. from eggs). The former should suffice to fix the `pkg_resources` problems with `metpy` (#5247), `branca` (#5256) and `folium` (#5262).

Fixes the "Providers" part of #4881.

The behavior of the new `PyiFrozenProvider` is modelled after `DefaultProvider` and `ZipProvider`. Wherever the behavior between the two is inconsistent, we implement the one that makes more sense for our case. The one deviation
in behavior is that in contrast to native providers, we cannot list the source `.py` files, as they do not exist in the frozen application.

The two new added tests, `test_pkg_resources_provider_source` and `test_pkg_resources_provider_frozen` verify the behavior of providers in original (source) and frozen application. They both use the same script with several behavior checks with test package in either package or egg form. The "source" test can be seen as a sort of a sanity check for the "frozen" one; by design of the test script and transitivity of its results, both tests passing indicate that `PyiFrozenProvider` implements behavior that is in line with the behavior of the two reference providers.

Legorooj commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+
+class _TocFilesystem:
+    def __init__(self, toc_files, toc_dirs=[]):
+        # Reconstruct the fileystem hierarchy by building a trie from
```
```suggestion
        # Reconstruct the fileystem hierarchy by building a tree from
```

rokm replied:
"Trie" is another name for "prefix tree", but I will replace it with the latter to avoid confusion.

Legorooj commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+
+        # ... as well as the actual one
+        if path.is_dir():
+            # Use os.listidr() to avoid having to convert Path objects
```
```suggestion
            # Use os.listdir() to avoid having to convert Path objects
```

rokm replied:
Ah, nicely spotted. Thanks!

Legorooj reviewed (commented):
It shows the quality of your work that the only problems I can find are typos! However just for safety's sake, I'll ask @bwoodsend to review this as well.

bwoodsend commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,92 @@
+
+    # Prepare the environment - default to 'os.environ'...
+    env = copy.deepcopy(os.environ)
+    # ... and prepend PYTHONPATH with pathex
+    if pathex:
+        if 'PYTHONPATH' in env:
+            pathex = os.pathsep.join([env.get('PYTHONPATH'), pathex])
```
In you're comment you say *prepend* but you actually *append*.

bwoodsend replied:
Also, any reason you chose not to use the `path_ex` option which should do the same thing? It's fine as it is. Initially I was concerned that defining `PYTHONPATH` would allow the built app to cheat and access the original package but it turn out it ignores `PYTHONPATH` at runtime.

rokm replied:
> In you're comment you say prepend but you actually append.

Nicely spotted, thanks! I was meaning to prepend it, although it probably makes no real difference.

> Also, any reason you chose not to use the path_ex option which should do the same thing? It's fine as it is. Initially I was concerned that defining PYTHONPATH would allow the built app to cheat and access the original package but it turn out it ignores PYTHONPATH at runtime.

The above function (`__exec_python_script`) is used only for the `source` test variant. The `frozen` variant is running without `PYTHONPATH` defined (there, `pathex` is passed to pyinstaller via `--paths` during the build).

htgoebel replied:
What about simplifying the code by making `pathex` mandatory and mandatory non-empty? As this is how is is used currently.

rokm replied:
> What about simplifying the code by making pathex mandatory and mandatory non-empty? As this is how is is used currently.

Indeed, that's what I'll be going for in the revised version. (As well as removal of unused `args`).

bwoodsend commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,92 @@
+from PyInstaller.utils.tests import importorskip
+from PyInstaller.compat import exec_python_rc
+
+# :todo: find a way to get this from `conftest` or such
```
I think we can tolerate this duplicity. If we do move the modules folder then the worst that can happen if we forget to update this is the tests will fail and we can correct it.

rokm replied:
OK, I'll remove the comment. It was carried over from an existing test...

bwoodsend commented:
I'd feel slightly more comfortable if those resource files weren't empty. When I do:

```python
>>> pkg_resources.resource_string("pyi_pkgres_testpkg", "subpkg1/data/extra/extra_entry1.txt")
b''
```

I question if something went wrong - even though I know you get a `FileNotFoundError` if something genuinely has gone wrong.

bwoodsend reviewed (approved):
I've been `code.interact()` messing about with this and it's very satisfyingly consistent with unfrozen Python.

rokm commented:
> I'd feel slightly more comfortable if those resource files weren't empty. When I do:
> 
> ```python
> >>> pkg_resources.resource_string("pyi_pkgres_testpkg", "subpkg1/data/extra/extra_entry1.txt")
> b''
> ```
> 
> I question if something went wrong - even though I know you get a `FileNotFoundError` if something genuinely has gone wrong.

Good point - I'll populate the data files in test package.

htgoebel reviewed (changes requested):
Impressive work! You've been so brave going down into the catacombs of pkg_resources. Wow!
Thanks for implementing this, this solves a major pain we have since long. And I like the idea of transparently including the files in the file-system.

Beside some nitpicking (see below and line comments), I have two relevant points:

**1)** Which takes precedense: PYZ or filesystem? This should be the same behavior as if a module is both in the PYZ and in the file-system (due to collect_data_files, including .pyc). I don't have the asnwer at hand, so please have a look at the code or test it.
Of course this needs to be documents, at least in the relase notes and at the top of the runt-time-hook..

**2)** AFAIU, `TocFilesystem` structure is a dict of dicts, containing `pathlib.PurePath`s as leafs. IMHO building this  is much too much overhead:
- Over the run-time of a program, resource access is rare and typically not done in a inner loop. This our Provider should not spend too much time, memory and eletrical power on speeding up the lookup of a resource. Chances are high that only a handful of resources will be requested even for a huge program.
  This is especially true since AFAIS, this is done for each call of `get_provider` (which is called again by echo `resource_*()` call), and not cached:
```
>>> p1 = get_provider("ansible")
>>> p2 = get_provider("ansible")
>>> id(p1)
140475825572560
>>> id(p2)
140475825511888
```
- While using pathlib here is elegant and simple, the overhead is huge. Again, esp. since results seem not to be cached, and the current code is touching many strings and files.
- Thus I suggest you have a look at ZipProvider and seem cant be imitated from there.

And now here are the minor remarks.

- Please put at least a `#` (and line-ending) into the empty files. Empyt file are easily deleted accidentally.
- Please choose different extensions for the data files, to avoid only .txt-files are collected ;-) If you want to stick with text-like file extensions, .rst and .md are good choices.
- Please provide a Makefile for recreating the .egg. Some distributions liek Debian want to have pure source code, and a zip file is a binary.
- Please use present tense in the commit messages.

htgoebel commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,91 @@
+    # ... and prepend PYTHONPATH with pathex
+    if pathex:
+        if 'PYTHONPATH' in env:
+            pathex = os.pathsep.join([pathex, env.get('PYTHONPATH')])
```
```suggestion
            pathex = os.pathsep.join([pathex, env.['PYTHONPATH']])
```
is already known to exist, so we ban safely use item-access.

rokm replied:
Indeed. Will be changed in next revision.

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+    def _has(self, path):
+        # Prevent access outside the package
+        path = self._normalize_path(path)
+        if not self._is_relative_to_package(path):
```
The `DefaultProvider` currently allows querying/listing of paths outside the given module/package via `..`, with an added deprecation warning: `<stdin>:1: DeprecationWarning: Use of .. or absolute path in a resource path is not allowed and will raise exceptions in a future release.`. 

`ZipProvider` cannot do it, and thus always returns `False` for `_has()` or `_isdir()`, and empty list for `_listdir()`.

In our `PyFrozenImplementation`, we explicitly disallow the jump, to keep behavior consistent between the PYZ-embedded and on-filesystem resources. In the filesystem case, the parent path could be queried otherwise. But in PYZ embeded case, the current implementation cannot query the parent path - even if the path was still within the "main" package (e.g., parent to a sub-package but still within the main one) because the prefix tree is constructed only from the base (sub)package down.

The `pyi_pkg_resources_provider.py` does contain tests for trying to query paths outside the base module/package to validate the behavior of `DefaultProvider`, `NullProvider`, and `PyiFrozenProvider`.

rokm commented:
Thanks for the detailed review!

> Impressive work! You've been so brave going down into the catacombs of pkg_resources. Wow!
> Thanks for implementing this, this solves a major pain we have since long. And I like the idea of transparently including the files in the file-system.

Perhaps it's worth pointing out that from the issues I've seen, we could probably get away by implementing only the on-filesystem part of the behavior here. 

If we aimed for that, it would probably be enough to just subclass the `DefaultProvider`, and override the  `_get()` method to use the original implementation from `NullProvider`. Tthe implementation from `DefaultProvider` is filesystem-only, while the one from `NullProvider` uses `loader.get_data()`, and thus in case of  `FrozenLoader` supports both on-filesystem and PYZ-embedded resources. 

That's the reason why the initial attempt to use `DefaultProvider` instead of `NullProvider` broke the existing `test_egg_zipped` test in `test_import.py`.

> Beside some nitpicking (see below and line comments), I have two relevant points:
> 
> **1)** Which takes precedense: PYZ or filesystem? This should be the same behavior as if a module is both in the PYZ and in the file-system (due to collect_data_files, including .pyc). I don't have the asnwer at hand, so please have a look at the code or test it.
> Of course this needs to be documents, at least in the relase notes and at the top of the runt-time-hook..

The PYZ-embedded resources take precedence. This is to keep behavior consistent with `FrozenImporter.get_data()`, which [also first checks the TOC and then filesystem](https://github.com/pyinstaller/pyinstaller/blob/04bb446f97e4c8ef32327d26a9c22c618cf79958/PyInstaller/loader/pyimod03_importers.py#L321-L342). (This method is used by `NullProvider.` and `PyiFrozenProvider._get()`, and in turn by `pkg_resources.resource_string()`).

> **2)** AFAIU, `TocFilesystem` structure is a dict of dicts, containing `pathlib.PurePath`s as leafs. IMHO building this is much too much overhead:
> 
>     * Over the run-time of a program, resource access is rare and typically not done in a inner loop. This our Provider should not spend too much time, memory and eletrical power on speeding up the lookup of a resource. Chances are high that only a handful of resources will be requested even for a huge program.
>       This is especially true since AFAIS, this is done for each call of `get_provider` (which is called again by echo `resource_*()` call), and not cached:
> 
> 
> ```
> >>> p1 = get_provider("ansible")
> >>> p2 = get_provider("ansible")
> >>> id(p1)
> 140475825572560
> >>> id(p2)
> 140475825511888
> ```

Ineed, `TocFilesystem` is a prefix tree implemented as dict of dicts, where leaf nodes should either:
* a string - last component of a file name
* empty dict - for indicating that the node is an empty directory; this is used to emulate PYZ-embedded sub-pacakge directories that originally contained only `.py` files and are now empty.

The prefix tree is there not so much to speed up the resource look-up, but more due to the fact that if we want to emulate PYZ-embedded filesystem, we need to somehow reconstruct the directory structure from filenames in the TOC. I could not come up with a better idea to do so - perhaps you have one?

But I agree with your performance concerns. Would you consider the following two modifications sufficient to alleviate those?

* Caching of prefix trees (in memory for the program run-time); i.e., a global dict with base package names for keys and TocFileystem structures for values. We could then also optimize this further to reuse the parts of base package trees for sub-packages and sub-modules, if necessary.

* The use of prefix tree is actually required only by the newly-implemented methods (`_has()`, `_isdir()` and `_listdir()`). We could defer the prefix tree creation (/retrieval from cache)  until the first call to one of these methods. This way, just creating the Provider without using any resource query/listing will incur no performance penalty.


>     * While using pathlib here is elegant and simple, the overhead is huge. Again, esp. since results seem not to be cached, and the current code is touching many strings and files.
> 
>     * Thus I suggest you have a look at ZipProvider and seem cant be imitated from there.

OK, I will take a closer look at `ZipProvider` implementation. If I recall correctly, it has split indices for directories and for files.  Would you prefer this design (over the current combined prefix tree9, or was your comment more in line with removing the use of `pathlib` in favor of `os.path`-based path manipulation?

> And now here are the minor remarks.
> 
>     * Please put at least a `#` (and line-ending) into the empty files. Empyt file are easily deleted accidentally.

Aha, good point. I filled in the datafiles as per an earlier comment, but the .py files are indeed blank. Will add a comment line in them.

>     * Please choose different extensions for the data files, to avoid only .txt-files are collected ;-) If you want to stick with text-like file extensions, .rst and .md are good choices.

Sure, we can diversify the extensions and contents of datafiles a bit. 

> 
>     * Please provide a Makefile for recreating the .egg. Some distributions liek Debian want to have pure source code, and a zip file is a binary.

Good point. I thought zipped eggs were allowed, since we already have `pyi_egg_zipped.egg` checked into the repository...

But on the other hand, keeping the "source" and zipped egg in sync is somewhat a hassle, so it might actually be better to have the test build the egg from the source and use it...

>     * Please use present tense in the commit messages.

Yeah, the commit messages are a bit of a mess right now, and reflect the actual incremental development history... I was hoping they would be squashed together with a  summary message from the PR. Otherwise, I will squash/reorganize them a bit once all other concerns are addresses and the PR is ready for merge...

rokm commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def __exec_python_script(script_filename, *args, pathex=None):
+    cmd = [script_filename]
+    cmd.extend(args)
```
Good point, thanks. But I'll simplify this a bit further, since no `args` are actually being passed.

rokm commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,91 @@
+    cmd.extend(args)
+
+    # Prepare the environment - default to 'os.environ'...
+    env = copy.deepcopy(os.environ)
```
Good point. Will be changed in next revision.

rokm commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def __get_test_package_path(package_type):
+    # Same test package, in three different formats
```
Ooops, yeah. Initially, there were three... will fix.

rokm commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,91 @@
+        if 'PYTHONPATH' in env:
+            pathex = os.pathsep.join([pathex, env.get('PYTHONPATH')])
+        env['PYTHONPATH'] = pathex
+
```
OK, will remove 'em.

rokm commented on tests/functional/test_pkg_resources_provider.py:
```diff
@@ -0,0 +1,91 @@
+    hooks_dir = os.path.join(_MODULES_DIR,
+                             'pyi_pkg_resources_provider',
+                             'hooks')
+
```
OK, will remove them for next revision.

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+#  * ZipProvider returns False
+#  > PyiFrozenProvider returns True
+ret = resource_exists(pkgname, '.')
+assert (is_default and ret is True) or \
+       (is_zip and ret is False) or \
+       (is_frozen and ret is True)
```
Hmmm, yes, but won't that hide (a hypothetical) case when the returned value is of non-boolean type? (And we would likely want to ensure that it is a boolean).

rokm replied:
On a second thought, we can probably afford not to be too paranoid about the returned value type. And the implicit comparison does look a lot cleaner.

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+# │       └── mod.py
+# └── subpkg3
+#     ├── _datafile.txt
+#     └── __init__.py
```
Good idea.

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+is_zip = isinstance(provider, ZipProvider)
+is_frozen = getattr(sys, 'frozen', False)
+
+assert any([is_default, is_zip, is_frozen]), "Unsupported provider type!"
```
Now that I see it, the suggest way is actually far more reasonable...

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+       (is_frozen and ret is True)
+
+# Package's directory, with empty path
+ret = resource_exists(pkgname, '')
+assert ret is True
```
Indeed, the intermediate assignment makes no sense in these simple cases. As for the explicit comparison against boolean True, see the comment above.

rokm replied:
As per the comment above, the explicit comparison to True/False has also been removed.

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+#  * ZipProvider returns False
+#  > PyiFrozenProvider disallows jumping to parent, and returns False
+# NOTE: using .. in path is deprecated and will raise exception in
+# future pkg_resources release
```
Will do. As far as I can tell from browsing old `setuptools` tags, the deprecation warning block first appeared in `setuptools` 40.8.0 (Feb 5, 2019).

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+
+if '__pycache__' in content:
+    content.remove('__pycache__')  # ignore __pycache__
+
```
OK, will do. 

I also think I will write up a helper function to minimize the code duplication among the content-listing tests...

rokm commented on tests/functional/scripts/pyi_pkg_resources_provider.py:
```diff
@@ -0,0 +1,514 @@
+if is_frozen:
+    expected = {x for x in expected if not x.endswith('.py')}
+
+content = resource_listdir(pkgname, '')
+content = set(content)
```
Good point. I think I'll need to revise the use of `set()` anyway, because it may be hiding the potential duplicated entries issue that you noted in another comment...

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
 # import hooks like FrozenImporter is. It uses method __loader__.get_data() in
 # methods pkg_resources.resource_string() and pkg_resources.resource_stream()
-res.register_loader_type(FrozenImporter, res.NullProvider)
+
```
OK.

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+# because it provides filesystem-only implementations (and overrides _get()
+# with a filesystem-only one), whereas our provider needs to also support
+# embedded resources.
+
```
OK.

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+
+        if not isinstance(node, dict):
+            return []  # Non-existant or file
+
```
OK, will do.

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+            return False
+
+        # Relative path for searching embedded resources
+        rel_path = path.relative_to(SYS_PREFIX)
+
+        return self.embedded_tree.path_exists(rel_path) or path.exists()
```
Ah, you're right. In this particular case the order of evaluations doesn't matter, so we can do a `path.exist()` check first and only if that fails proceed with the `rel_path` computation and check in the embedded tree.

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+        if path.is_dir():
+            # Use os.listdir() to avoid having to convert Path objects
+            # to strings...
+            path = str(path)  # python 3.5
```
Good idea.

rokm commented on PyInstaller/hooks/rthooks/pyi_rth_pkgres.py:
```diff
@@ -9,14 +9,179 @@
+            # Use os.listdir() to avoid having to convert Path objects
+            # to strings...
+            path = str(path)  # python 3.5
+            content += os.listdir(path)
```
Hmm, indeed. And I suspect this may already be happening for some of the directories, but we're missing that in the test because `set()` is used to compare the results. Will fix both...

Legorooj commented:
@rokm don't bother with fixing commit messages yet please - get the actual code work done, then we can fix commit messages with a rebase.

rokm commented on news/5284.hooks.rst:
```diff
@@ -0,0 +1 @@
+Implement package content listing via ``pkg_resources``. Add a custom provider for ``FrozenImporter`` that enables support for ``pkg_resources.resource_exists()``, ``pkg_resources.resource_isdir()``, and ``pkg_resources.resource_listdir()``.
```
Done.

htgoebel commented on news/5284.hooks.rst:
```diff
@@ -0,0 +1,5 @@
+Add support for package content listing via ``pkg_resources`` by
+implementing a custom ``PyiFrozenProvider`` for ``FrozenImporter``. The
```
```suggestion
Add support for package content listing via ``pkg_resources``. The
```

rokm replied:
Done.

htgoebel commented on tests/functional/modules/pyi_pkg_resources_provider/package/pyi_pkgres_testpkg/a.py:
```diff
@@ -0,0 +1,10 @@
+# The full license is in the file COPYING.txt, distributed with this software.
+#
+# SPDX-License-Identifier: (GPL-2.0-or-later WITH Bootloader-exception)
+#-----------------------------------------------------------------------------
```
Let's save electric power and bandwidth: These are empty files, no need for a copyright header. A `#`suffices.

Same for the other empty ones.

rokm replied:
OK, makes sense.

htgoebel reviewed (changes requested):
Beside some minor nit-picking, this looks good for me. Okay to merge after this have been done and commits cleaned up.

(I would have applied these nit-picking changes myself, but commits are not yet clean up.)

rokm commented:
Alright, here's the updated PR, rebased on top of `develop`.

I've squashed all test-related commits into first commit, and most initial incremental work on the `PyFrozenProvider` into second one. Couple of larger subsequent modifications are still in separate commits, but we can squash them all into the second one, if that's preferable.
